### PR TITLE
feat!: cache get params for 200 blocks

### PR DESCRIPTION
**Pull request by RafilxTenfen**
## Description
- Add cache for [price feeder](https://github.com/umee-network/umee/blob/e13cb3e9939e6903b9ba2cadbd9ea21f5217fb7c/price-feeder/oracle/oracle.go#L282) `o.GetParams` for 200 blocks (hardcoded)

closes: #1049 

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] added appropriate labels to the PR
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

**Review by brianosaurus — APPROVED** [review]
:+1: Nice work!

**Review by dreamcodez — APPROVED** [review]
looks good to me; want to hear @adamewozniak on this as well

File: price-feeder/CHANGELOG.md [hunk 1]
```diff
@@ -49,6 +49,7 @@ Ref: https://keepachangelog.com/en/1.0.0/
 
 - [#1032](https://github.com/umee-network/umee/pull/1032) Update the accepted tvwap period from 3 minutes to 5 minutes.
 - [#978](https://github.com/umee-network/umee/pull/978) Cleanup the oracle package by moving deviation & conversion logic.
+- [#1050](https://github.com/umee-network/umee/pull/1050) Cache x/oracle params to decrease the number of queries to node.
```
Comment (adamewozniak) [inline price-feeder/CHANGELOG.md:52]:
```suggestion
- [#1050](https://github.com/umee-network/umee/pull/1050) Cache x/oracle params to decrease the number of queries to nodes.
```

File: price-feeder/oracle/oracle.go [hunk 2]
```diff
@@ -31,7 +31,11 @@ import (
 // and broadcast pre-vote and vote transactions such that they're committed in
 // at least one block during each voting period.
 const (
-	tickerSleep = 1000 * time.Millisecond
+	tickerSleep = 1500 * time.Millisecond
```
Comment (adamewozniak) [inline price-feeder/oracle/oracle.go:34]: Let's not change this - we'd be using less node queries, but we'd miss more voting periods.
```suggestion
	tickerSleep = 1000 * time.Millisecond
```

File: price-feeder/oracle/oracle.go [hunk 3]
```diff
@@ -31,7 +31,11 @@ import (
...
+
+	// cacheOnChainBlockQuantity represents the amount of blocks
+	// in which the oracle onchain dat(a will wait to be updated
+	cacheOnChainBlockQuantity = uint64(200)
```
Comment (adamewozniak) [inline price-feeder/oracle/oracle.go:38]:
```suggestion
	// paramsCacheInterval represents the amount of blocks
	// during which we will cache the oracle params.
	paramsCacheInterval = uint64(200)
```

File: price-feeder/oracle/oracle.go [hunk 4]
```diff
@@ -69,6 +73,15 @@ type Oracle struct {
 	mtx             sync.RWMutex
 	lastPriceSyncTS time.Time
 	prices          map[string]sdk.Dec
+
```
Comment (adamewozniak) [inline price-feeder/oracle/oracle.go:76]:
```suggestion
```

File: price-feeder/oracle/oracle.go [hunk 5]
```diff
@@ -69,6 +73,15 @@ type Oracle struct {
...
+	oracleOnChain OnChainData
+}
+
+// OnChainData used to cache data that can be updated
+// at each cacheOnChainBlockQuantity amount of blocks
+type OnChainData struct {
+	params           *oracletypes.Params
+	lastUpdatedBlock uint64
```
Comment (adamewozniak) [inline price-feeder/oracle/oracle.go:84]:
```suggestion
// ParamCache is used to cache oracle param data for
// an amount of blocks, defined by paramsCacheInterval.
type ParamCache struct {
	params           *oracletypes.Params
	lastUpdatedBlock int64
```

File: price-feeder/oracle/oracle.go [hunk 6]
```diff
@@ -99,6 +112,7 @@ func New(
 		previousPrevote: nil,
 		providerTimeout: providerTimeout,
 		deviations:      deviations,
+		oracleOnChain:   OnChainData{},
```
Comment (adamewozniak) [inline price-feeder/oracle/oracle.go:115]:
```suggestion
		paramCache:   ParamCache{},
```

File: price-feeder/oracle/oracle.go [hunk 7]
```diff
@@ -166,7 +180,7 @@ func (o *Oracle) GetPrices() map[string]sdk.Dec {
 // to determine prices. If candles are not available, uses the most recent prices
 // with VWAP. Warns the the user of any missing prices, and filters out any faulty
 // providers which do not report prices or candles within 2𝜎 of the others.
-func (o *Oracle) SetPrices(ctx context.Context, acceptList oracletypes.DenomList) error {
+func (o *Oracle) SetPrices(ctx context.Context) error {
```
Comment (adamewozniak) [inline price-feeder/oracle/oracle.go:183]: Cleaner 💯 nice

File: price-feeder/oracle/oracle.go [hunk 8]
```diff
@@ -460,22 +522,22 @@ func (o *Oracle) tick(ctx context.Context) error {
...
-	if err := o.SetPrices(ctx, oracleParams.AcceptList); err != nil {
-		return err
+	if blockHeight < 1 {
+		return fmt.Errorf("expected positive block height")
 	}
 
-	o.checkAcceptList(oracleParams)
-
-	blockHeight, err := rpcclient.GetChainHeight(clientCtx)
+	oracleParams, err := o.GetCachedParams(uint64(blockHeight))
```
Comment (adamewozniak) [inline price-feeder/oracle/oracle.go:534]: Why don't we store cached block height as an int64? that way we don't have to convert anything here
```suggestion
	oracleParams, err := o.GetCachedParams(blockHeight)
```

File: price-feeder/oracle/oracle.go [hunk 9]
```diff
@@ -356,6 +370,54 @@ func SetProviderTickerPricesAndCandles(
...
 }
 
+// Update update the values inside the OnChainData struct.
+func (onChainData *OnChainData) Update(currentBlockHeigh uint64, params oracletypes.Params) {
+	onChainData.lastUpdatedBlock = currentBlockHeigh
+	onChainData.params = &params
+}
+
+func (onChainData *OnChainData) IsParamsOutdated(currentBlockHeigh uint64) bool {
+	// doesn't have any data
```
Comment (adamewozniak) [inline price-feeder/oracle/oracle.go:380]:
```suggestion
```

File: price-feeder/oracle/oracle.go [hunk 10]
```diff
@@ -356,6 +370,54 @@ func SetProviderTickerPricesAndCandles(
...
+	// state in which the blockchain is under cacheOnChainBlockQuantity
+	// amount of blocks, but already has value for params
+	if currentBlockHeigh < cacheOnChainBlockQuantity {
+		return false
+	}
+
+	// an edge case, should never happen
+	// which the current blockchain height is lower
+	// than the last updated block, to fix should
+	// just update the on chain params again
```
Comment (adamewozniak) [inline price-feeder/oracle/oracle.go:394]: When does this happen? Are you just making sure we're covering all code paths? 

```suggestion
	// This is an edge case, which should never happen.
	// The current blockchain height is lower
	// than the last updated block, to fix we should
	// just update the cached params again.
```

File: price-feeder/oracle/oracle.go [hunk 11]
```diff
@@ -356,6 +370,54 @@ func SetProviderTickerPricesAndCandles(
...
+		return true
+	}
+
+	return (currentBlockHeigh - onChainData.lastUpdatedBlock) > cacheOnChainBlockQuantity
+}
+
+// GetCachedParams returns the last updated parameters of the x/oracle module
+// if the current oracleOnChainParams is outdated we will query it again to
+// get the current oracle params.
+func (o *Oracle) GetCachedParams(currentBlockHeigh uint64) (oracletypes.Params, error) {
```
Comment (adamewozniak) [inline price-feeder/oracle/oracle.go:405]:
```suggestion
// GetParamCache returns the last updated parameters of the x/oracle module
// if the current ParamCache is outdated, we will query it again.
func (o *Oracle) GetParamCache(currentBlockHeigh uint64) (oracletypes.Params, error) {
```

File: price-feeder/oracle/oracle.go [hunk 12]
```diff
@@ -356,6 +370,54 @@ func SetProviderTickerPricesAndCandles(
...
+	if !o.oracleOnChain.IsParamsOutdated(currentBlockHeigh) {
+		return *o.oracleOnChain.params, nil
```
Comment (adamewozniak) [inline price-feeder/oracle/oracle.go:407]: Why do we use a pointer here? Can we just remove all the reference & dereferencers?

File: price-feeder/oracle/oracle_test.go [hunk 13]
```diff
@@ -351,6 +343,63 @@ func (ots *OracleTestSuite) TestPrices() {
 	ots.Require().Equal(sdk.MustNewDecFromStr("1"), prices["USDT"])
 }
 
+func TestOnChainDataIsParamsOutdated(t *testing.T) {
```
Comment (adamewozniak) [inline price-feeder/oracle/oracle_test.go:346]: Awesome 🚀

File: price-feeder/oracle/oracle.go [hunk 14]
```diff
@@ -69,6 +73,15 @@ type Oracle struct {
 	mtx             sync.RWMutex
 	lastPriceSyncTS time.Time
 	prices          map[string]sdk.Dec
+
+	oracleOnChain OnChainData
+}
+
+// OnChainData used to cache data that can be updated
+// at each cacheOnChainBlockQuantity amount of blocks
+type OnChainData struct {
```
Comment (adamewozniak) [inline price-feeder/oracle/oracle.go:82]: Can we move this struct, associated functions, and variables to another small file? Trying to keep the `oracle.go` file as small as we can

**Review by adamewozniak — CHANGES_REQUESTED** [review]
Mostly just nitpics on naming & comments - but all good 🚀 

Main thing I'd like to keep from changing is the ticker timeout here

File: price-feeder/oracle/oracle.go (hunk 14 above)
Comment (RafilxTenfen) [reply price-feeder/oracle/oracle.go:82]: sure thing

File: price-feeder/oracle/oracle.go (hunk 12 above)
Comment (RafilxTenfen) [reply price-feeder/oracle/oracle.go:407]: was to be able to check if `o.oracleOnChain.params == nil`

File: price-feeder/oracle/oracle.go (hunk 8 above)
Comment (RafilxTenfen) [reply price-feeder/oracle/oracle.go:530]: to be able to use [here](https://github.com/umee-network/umee/blob/c5a846e4a6ccf7646532d4313d4a9274f6884cd3/price-feeder/oracle/oracle.go#L399)

Comment (adamewozniak) [reply price-feeder/oracle/oracle.go]: You should just be able to use o.oracleOnChain.params == oracletypes.Params{} 👍

File: price-feeder/oracle/param.go [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+// ParamCache is used to cache oracle param data for
+// an amount of blocks, defined by paramsCacheInterval.
+type ParamCache struct {
+	params           *oracletypes.Params
+	lastUpdatedBlock int64
+}
+
+// Update retrieves the most recent oracle params and
+// updates the instance.
+func (onChainData *ParamCache) Update(currentBlockHeigh int64, params oracletypes.Params) {
```
Comment (adamewozniak) [inline price-feeder/oracle/param.go:20]:
```suggestion
func (paramCache *ParamCache) Update(currentBlockHeigh int64, params oracletypes.Params) {
```

File: price-feeder/oracle/param.go [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+	onChainData.lastUpdatedBlock = currentBlockHeigh
+	onChainData.params = &params
+}
+
+// IsParamsOutdated checks whether or not the current
+// param data was fetched in the last 200 blocks.
+func (onChainData *ParamCache) IsParamsOutdated(currentBlockHeigh int64) bool {
```
Comment (adamewozniak) [inline price-feeder/oracle/param.go:27]:
```suggestion
func (paramCache *ParamCache) IsOutdated(currentBlockHeigh int64) bool {
```

File: price-feeder/oracle/oracle.go (hunk 12 above)
Comment (RafilxTenfen) [reply price-feeder/oracle/oracle.go]: image
Go didn't liked

File: price-feeder/oracle/param.go (hunk 16 above)
Comment (RafilxTenfen) [reply price-feeder/oracle/param.go]: good catch

File: price-feeder/oracle/oracle.go (hunk 12 above)
Comment (RafilxTenfen) [reply price-feeder/oracle/oracle.go]: What do you think about using this way? 
image

Comment (RafilxTenfen) [reply price-feeder/oracle/oracle.go]: @adamewozniak I am going to ship as pointer right now

**Pull request merged**